### PR TITLE
Auth/api intregation

### PR DESCRIPTION
**Pull request by GCham5**
please provide feedback.
Note: there might be some unused/commented out code

### CODE REVIEW

File: AuthAPI/Services/AuthMessageSender.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace TreasuryBoardSecretariat.AuthAPI.Services
+{
+    public class AuthMessageSender
+    {
+        public string SendGridUser { get; set; }
+        public string SendGridKey { get; set; }
+    }
```
Comment (Thomas-Langlais) [inline AuthAPI/Services/AuthMessageSender.cs]: Is this for 2FA?

File: AuthAPI/Services/EmailSender.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            return Execute(Options.SendGridKey, subject, message, email);
+        }
+
+        public Task Execute(string apiKey, string subject, string message, string email)
+        {
+            var client = new SendGridClient(apiKey);
+            var msg = new SendGridMessage()
+            {
+                From = new EmailAddress("gcam042@uottawa.ca", Options.SendGridUser),
```
Comment (Thomas-Langlais) [inline AuthAPI/Services/EmailSender.cs]: The From attribute should be configured in the appsettings.*.json file

File: webapp/Startup.cs [hunk 3]
```diff
@@ -23,6 +26,24 @@ public Startup(IConfiguration configuration)
...
+
+            services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme).AddJwtBearer(options =>
+            {
+                options.TokenValidationParameters = new TokenValidationParameters
+                {
+                    ValidateIssuer = true,
+                    ValidateAudience = true,
+                    ValidateIssuerSigningKey = true,
+                    ValidIssuer = "mysite.com",
+                    ValidAudience = "mysite.com",
```
Comment (Thomas-Langlais) [inline webapp/Startup.cs]: Same thing, this shouldn't be configured here, it should be in the appsettings file

File: webapp/Views/Home/Index.cshtml [hunk 4]
```diff
@@ -84,7 +84,7 @@
         <h2>CORTEX Prototypes</h2>
         <ul>
             <li>About CORTEX Prototypes</li>
-            <li>Why <a href="https://en.wikipedia.org/wiki/Software_prototyping">Prototypes?</a></li>
+            <li>Why <a hre f="https://en.wikipedia.org/wiki/Software_prototyping">Prototypes?</a></li>
```
Comment (Thomas-Langlais) [inline webapp/Views/Home/Index.cshtml]: mistake

File: webapp/Views/Home/Index.cshtml [hunk 5]
```diff
@@ -118,4 +118,5 @@
 
 @section Scripts {
     <script src="~/js/login/site.js"></script>
+    <script src="~/js/site.js"></script>
```
Comment (Thomas-Langlais) [inline webapp/Views/Home/Index.cshtml]: This is added in the layout file

File: webapp/Views/Shared/_Layout.cshtml [hunk 6]
```diff
@@ -106,24 +106,28 @@
...
+</environment>
     <environment exclude="Development">
         <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
                 asp-fallback-src="~/lib/jquery/dist/jquery.min.js"
                 asp-fallback-test="window.jQuery"
                 crossorigin="anonymous"
                 integrity="sha384-K+ctZQ+LL8q6tP7I94W+qzQsfRV2a+AfHIi9k8z8l9ggpc8X+Ytst4yBo/hH+8Fk">
         </script>
+
+
```
Comment (Thomas-Langlais) [inline webapp/Views/Shared/_Layout.cshtml]: remove the spaces

File: webapp/Views/Shared/_LoginPartial.cshtml [hunk 7]
```diff
@@ -1,22 +1,25 @@
...
+            {
+                <!--<li>
+                    <a asp-area="" asp-controller="Account" asp-action="RegisterEmployee" title="Register">Register a new employee (Not Implemented)</a>
+                </li>-->
+            }
+            <li>
+                <a href="javascript:{}" onclick="document.getElementById('logoutForm').submit();">Log out</a>
+            </li>
+        </ul>
+    </form> *@
```
Comment (Thomas-Langlais) [inline webapp/Views/Shared/_LoginPartial.cshtml]: Remove the comment

File: webapp/wwwroot/js/account/login.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    console.log(history.back);
+}
+
+loginBtn.addEventListener("click", function () {
+    if (ValidateForm()) {
+        loginCall();
+    }
+});
+
+function ValidateForm() {
```
Comment (Thomas-Langlais) [inline webapp/wwwroot/js/account/login.js]: This should comply with es6 naming conventions

File: webapp/wwwroot/js/account/login.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        alert("Please fill out all the fields");
+        return false;
+    } else {
+        return true;
+    }
+}
+
+const loginCall = async () => {
+
+    let resp = await fetch(`https://localhost:44363/api/auth/login`, {
```
Comment (Thomas-Langlais) [inline webapp/wwwroot/js/account/login.js]: We will need to change this in the future considering that the api domain will want to point to the azure cloud instance

File: webapp/wwwroot/js/account/register.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
+﻿var regBtn = document.getElementById("regBtn");
+
+var email = document.getElementById("email");
+
+var password = document.getElementById("password");
+
+var confirmPassword = document.getElementById("confirmPassword");
+
+regBtn.addEventListener("click", function () {
+    if (ValidateForm()) {
```
Comment (Thomas-Langlais) [inline webapp/wwwroot/js/account/register.js]: Same comment as above

File: webapp/wwwroot/js/account/register.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        registerCall();
+    }
+});
+
+
+const registerCall = async () => {
+    let resp = await fetch(`https://localhost:44363/api/auth/register`, {
```
Comment (Thomas-Langlais) [inline webapp/wwwroot/js/account/register.js]: same thing as above

File: webapp/wwwroot/js/home/index.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+var aToken = localStorage.getItem("token");
+var rToken = localStorage.getItem("refresh");
+
+logoutBtn.addEventListener('click', function () {
+    //setting to X instead of removing
+    //reason: mistakenly did not check for null and might take a bit of work 
+    localStorage.removeItem('token');
+    localStorage.removeItem('refresh');
+    window.location.reload(true);
+    window.location = 'https://localhost:44375/';
```
Comment (Thomas-Langlais) [inline webapp/wwwroot/js/home/index.js]: We shouldn't point to localhost, we need to setup a way to choose the domain on startup of the webapp.

File: webapp/wwwroot/js/pip/index.js [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+    var aToken = localStorage.getItem("token");
+    var rToken = localStorage.getItem("refresh");
+
+
+    if (isExpired(aToken, rToken)) {
+        // fetch from vaultAPI
+    } else {
+        alert("please login");
+        window.location = "https://localhost:44375/Account/Login";
```
Comment (Thomas-Langlais) [inline webapp/wwwroot/js/pip/index.js]: same comment with hardcoding the domain

File: webapp/wwwroot/js/tokenValidation.js [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+        return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
+    }).join(''));
+
+    return JSON.parse(jsonPayload);
+};
+
+//function to get new access token
+async function getNewAccessToken(rToken) {
+    const data = { refreshToken: rToken };
+    let resp = await fetch(`https://localhost:44363/api/auth/refresh`, {
```
Comment (Thomas-Langlais) [inline webapp/wwwroot/js/tokenValidation.js]: same domain issue

File: webapp/wwwroot/js/tokenValidation.js [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+
+//function to parse JWT
+function parseJwt(token) {
+    var base64Url = token.split('.')[1];
+    var base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
+    var jsonPayload = decodeURIComponent(atob(base64).split('').map(function (c) {
+        return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
+    }).join(''));
```
Comment (Thomas-Langlais) [inline webapp/wwwroot/js/tokenValidation.js]: What does this function do?
To parse the jwt, I thought the only thing you had to do was the following.

```js
let jwtObj = JSON.parse(atob(token.split('.')[1]));
```

File: webapp/Models/Account/LoginViewModel.cs [hunk 16]
```diff
@@ -0,0 +1,18 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace webapp.Models.Account
+{
+    public class LoginViewModel
```
Comment (Thomas-Langlais) [inline webapp/Models/Account/LoginViewModel.cs]: How is the Register and Login view used?
Isn't the button have a custom click handler?

File: AuthAPI/Controllers/AuthController.cs [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class AuthController : Controller
+    {
+        private readonly IConfiguration _configuration;
+        private readonly UsersContext _context;
+        private readonly UserManager<IdentityUser> _userManager;
+        private readonly SignInManager<IdentityUser> _signInManager;
+        private readonly IOptions<AuthMessageSender> _authMessageSender;
```
[the remark this answers is not in the document]
Comment (GCham5) [reply AuthAPI/Controllers/AuthController.cs]: I read the WIKI article however this [article](https://github.com/dotnet/corefx/blob/368fdfd86ee3a3bf1bca2a6c339ee590f3d6505d/Documentation/coding-guidelines/coding-style.md) uses underscores. What do you recommend? (I might say sticking with underscores as seems to be the more recent convention)

[the remark this answers is not in the document]
Comment (Thomas-Langlais) [reply AuthAPI/Controllers/AuthController.cs]: While I agree with that article being an up to date source and appreciate you sticking with the convention, but this is the convention that CoreFX uses for its libraries. We define our own convention to use and follow ourselves. In this case I chose to reference the naming convention to be the official Microsoft naming convention uses.

If you can find in the link in the wiki about naming private variables this way, then that is fine with me.

File: AuthAPI/Controllers/AuthController.cs [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+            IPrincipal principal = tokenHandler.ValidateToken(refreshToken, validationParameters, out validatedToken);
+            return true;
+        }
+
+        private TokenValidationParameters GetValidationParameters()
+        {
+            return new TokenValidationParameters
+            {
+                ValidateIssuer = false,
+                ValidateAudience = false,
```
[the remark this answers is not in the document]
Comment (GCham5) [reply AuthAPI/Controllers/AuthController.cs]: I made them false as I did not include issuer or audience in the refresh tokens (testing purposes). However, I can easily add those

File: AuthAPI/Services/AuthMessageSender.cs (hunk 1 above)
Comment (GCham5) [reply AuthAPI/Services/AuthMessageSender.cs]: Yes, however, I placed a pause on 2FA as we discussed before. AuthMessageSender, EmailSender, and IEmailSender are for 2FA

File: webapp/Models/Account/LoginViewModel.cs (hunk 16 above)
Comment (GCham5) [reply webapp/Models/Account/LoginViewModel.cs]: You're right, no use. Initially had them when I was implementing login and register (before using js). At the moment, their only use is for validation but I will do that in js. I'll remove them.

File: webapp/wwwroot/js/home/index.js (hunk 12 above)
Comment (GCham5) [reply webapp/wwwroot/js/home/index.js]: atm i have implemented this solution: 
` window.location = https://${window.location.host};`

File: webapp/wwwroot/js/pip/index.js (hunk 13 above)
Comment (GCham5) [reply webapp/wwwroot/js/pip/index.js]: I believe the fix I mentioned above could work if we keep the URL paths the same

File: webapp/wwwroot/js/tokenValidation.js (hunk 15 above)
Comment (GCham5) [reply webapp/wwwroot/js/tokenValidation.js]: you're right. changed it, thanks

**Pull request merged**